drivers/mspi/mspi_stm32_xspi.c

```diff
@@ -0,0 +1,1899 @@
+ *
+ */
+static int read_write_in_memory_map_mode(const struct device *dev, struct mspi_xfer_packet *packet)
+{
+	int ret;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (packet->data_buf == NULL) {
+		LOG_ERR("data buf is null :%x", packet->cmd);
+		return -EIO;
+	}
+
+	if (!mspi_stm32_xspi_is_memorymap(dev)) {
+		ret = mspi_stm32_xspi_memmap_on(dev);
+		if (ret != 0) {
+			LOG_ERR("Failed to set memory mapped");
+			return ret;
+		}
+	}
+
+	uintptr_t mmap_addr = dev_data->memmap_base_addr + packet->address;
+
+	if (packet->dir == MSPI_RX) {
+		LOG_INF("Memory-mapped read from 0x%08lx, len %zu", mmap_addr, packet->num_bytes);
+		memcpy(packet->data_buf, (void *)mmap_addr, packet->num_bytes);
+		k_sleep(K_MSEC(1));
+		return 0;
+	}
+#if defined(CONFIG_MSPI_STM32_ALLOW_WRITE_IN_MEMMAP)
+	LOG_INF("Memory-mapped write from 0x%08lx, len %zu", mmap_addr, packet->num_bytes);
+	memcpy((void *)mmap_addr, packet->data_buf, packet->num_bytes);
+	k_sleep(K_MSEC(1));
+	return 0;
+#else
+
+	ret = mspi_stm32_xspi_abort_memmap_if_enabled(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return -EPROTONOSUPPORT;
+#endif
+}
+static HAL_StatusTypeDef read_write_in_indirect_mode(const struct device *dev,
+						     struct mspi_xfer_packet *packet,
+						     uint8_t access_mode)
+{
+	HAL_StatusTypeDef hal_ret;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (packet->dir == MSPI_RX) {
+		/* Receive the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			hal_ret = HAL_XSPI_Receive(&dev_data->hmspi.xspi, packet->data_buf,
+						   HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+			goto end;
+		case MSPI_ACCESS_ASYNC:
+			hal_ret = HAL_XSPI_Receive_IT(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			uint8_t *dma_buf = k_aligned_alloc(CONFIG_MSPI_STM32_BUFFER_ALIGNMENT,
+							   packet->num_bytes);
+
+			if (dma_buf == NULL) {
+				LOG_ERR("DMA buffer allocation failed");
+				return HAL_ERROR;
+			}
+
+			hal_ret = HAL_XSPI_Receive_DMA(&dev_data->hmspi.xspi, dma_buf);
+			if (hal_ret == HAL_OK) {
+				if (k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
+					LOG_ERR("Failed to take sem");
+					k_free(dma_buf);
+					return HAL_BUSY;
+				}
+				memcpy(packet->data_buf, dma_buf, packet->num_bytes);
+			}
+
+			k_free(dma_buf);
+			goto end;
+		default:
+			/* Not correct */
+			hal_ret = HAL_BUSY;
+			break;
+		}
+	} else {
+		/* Transmit the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			hal_ret = HAL_XSPI_Transmit(&dev_data->hmspi.xspi, packet->data_buf,
+						    HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+			goto end;
+		case MSPI_ACCESS_ASYNC:
+			hal_ret = HAL_XSPI_Transmit_IT(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			hal_ret = HAL_XSPI_Transmit_DMA(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		default:
+			/* Not correct */
+			hal_ret = HAL_BUSY;
+			break;
+		}
+	}
+
+	if ((hal_ret != HAL_OK) || (access_mode == MSPI_ACCESS_SYNC)) {
```

Comment:
`access_mode == MSPI_ACCESS_SYNC` cases (above) already use `goto end` instructions.
```suggestion
	if (hal_ret != HAL_OK) {
```
or
```c
	/* Upon success, lock again expecting the IRQ for end of Tx or Rx */
	if (hal_ret == HAL_OK && k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
		LOG_ERR("Failed to take semaphore");
		return HAL_BUSY;
	}
```